File: apps/macos/Conformance.swift
import Foundation

@main
struct Conformance {
    static func main() throws {
        let args = CommandLine.arguments
        guard (4...6).contains(args.count) else { fatalError("usage: conformance INPUT LAYER_ID OUTPUT [PATCH_FILE] [image]") }
        let session = try NativeSession.open(source: String(contentsOfFile: args[1], encoding: .utf8))
        let opened = try session.exportJson()
        if args.count == 6 && args[5] == "image" {
            _ = try session.setImage(layerId: args[2], patchJson: String(contentsOfFile: args[4], encoding: .utf8))
        } else if args.count == 5 {
            _ = try session.setText(layerId: args[2], patchJson: String(contentsOfFile: args[4], encoding: .utf8))
        } else {
            _ = try session.setScanlines(layerId: args[2], amount: 50)
        }
        let changed = try session.exportJson()
        let didUndo = try session.undo()
        precondition(didUndo)
        let undone = try session.exportJson()
        let didRedo = try session.redo()
        precondition(didRedo)
        let redone = try session.exportJson()
        let reopened = try NativeSession.open(source: redone).exportJson()
        let output = ["opened": opened, "changed": changed, "undone": undone, "redone": redone, "reopened": reopened]
        try JSONEncoder().encode(output).write(to: URL(fileURLWithPath: args[3]), options: .atomic)
        print("Swift/UniFFI command sequence complete")
    }
}

File: crates/artifact-core/examples/conformance.rs
use artifact_core::DocumentSession;
use serde_json::json;
use std::{env, fs};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = env::args().collect();
    if !(4..=6).contains(&args.len()) {
        return Err("usage: conformance INPUT LAYER_ID OUTPUT [PATCH_FILE] [image]".into());
    }
    let mut session = DocumentSession::open(&fs::read_to_string(&args[1])?)?;
    let opened = session.export_json();
    let mut steps = 1;
    if args.len() == 6 && args[5] == "editor" {
        let commands: Vec<serde_json::Value> =
            serde_json::from_str(&fs::read_to_string(&args[4])?)?;
        steps = 0;
        for command in commands {
            if session.execute(&command.to_string())? {
                steps += 1;
            }
        }
    } else if args.len() == 6 && args[5] == "image" {
        session.set_image(&args[2], &fs::read_to_string(&args[4])?)?;
    } else if args.len() == 5 {
        session.set_text(&args[2], &fs::read_to_string(&args[4])?)?;
    } else {
        session.set_scanlines(&args[2], 50.0)?;
    }
    let changed = session.export_json();
    for _ in 0..steps {
        assert!(session.undo());
    }
    let undone = session.export_json();
    for _ in 0..steps {
        assert!(session.redo());
    }
    let redone = session.export_json();
    let reopened = DocumentSession::open(&redone)?.export_json();
    fs::write(&args[3], json!({"opened": opened, "changed": changed, "undone": undone, "redone": redone, "reopened": reopened}).to_string())?;
    println!("Native Rust command sequence complete");
    Ok(())
}
